### ISSUE TITLE
Feature idea: BatchUpdateBody for easier configuration of update BatchOperations

### ISSUE THREAD

**Issue by AndyHaskell** [algolia/algoliasearch-client-go#310]
Right now if I want to do an `AddUnique` on a field I'm updating in a `BatchOperation`, I'd need to make a field that marshals to JSON as
```
{
    "_operation": "AddUnique",
    "value": "the item I'm adding"
}
```

So I had an idea for this type to streamline writing more intricate batch operations

```go
type BatchOpBody map[string]interface{}

type FieldAction struct {
        Op    `json:"_operation"`
        Value `json:"value"`
}

func (b BatchOpBody) SetFieldAction(field, action string) {
        value := b[field]
        b[field] = FieldAction{
                Op: action,
                Value: value,
        }
}
```

Then someone updating one field, adding an item to a second field, and removing an item from a third field could do something like:

```go
b := algoliasearch.BatchOpBody{
        "fieldImUpdating":      100,
        "fieldImAppendingTo":  "item I'm adding",
        "fieldImRemovingFrom": "item I'm removing", 
}
b.SetFieldAction("fieldImAppendingTo", "AddUnique")
b.SetFieldAction("fieldImRemovingFrom", "Remove")

op := algoliasearch.BatchOperation{
        "Action": "partialUpdateObject",
        "Body": b,
}
```

What do you think?

**Comment by aseure** [comment]
Hello @AndyHaskell 

Thank you for opening this issue. I've just tried the `BatchOperation` by myself with a `partialUpdateObject` and I agree, it's not convenient and we need to do something to make it easier.

I like your proposal. However, I still find it a little bit difficult to use as you need to:
 - type the name of the attribute two times
 - let the user put the exact operation name

To help with those issues, what do you think of this proposal instead: https://github.com/algolia/algoliasearch-client-go/pull/314?

Thanks to this, the example you have written may be rewritten like this:
```go
batchOps := []algoliasearch.BatchOperation{
	{
		Action: "partialUpdateObject",
		Body: map[string]interface{}{
			"objectID":            "2000783612",
			"fieldImUpdating":     100,
			"fieldImAppendingTo":  algoliasearch.AddUniqueOp("item I'm adding"),
			"fieldImRemovingFrom": algoliasearch.RemoveOp("item I'm removing"),
		},
	},
}

res, err := index.Batch(batchOps)
```

I've just tried it against a real Algolia index and it correctly updates the following record:
```
{
  "fieldImUpdating": 50,
  "fieldImAppendingTo": [
    "first object"
  ],
  "fieldImRemovingFrom": [
    "first object",
    "item I'm removing"
  ],
  "objectID": "2000783612"
}
```

into:
```json
{
  "fieldImUpdating": 100,
  "fieldImAppendingTo": [
    "first object",
    "item I'm adding"
  ],
  "fieldImRemovingFrom": [
    "first object"
  ],
  "objectID": "2000783612"
}
```

What do you think?

**Comment by AndyHaskell** [comment]
Nice! I like that

**Comment by aseure** [comment]
Ok I'll add the tests for that on the PR and merge/tag it.

**Comment by aseure** [comment]
Functions now available as of version `2.11.0`. Have a good week-end and thank you again for contributing.
